Split Model to have BaseModel without database related stuff

najdanovicivan opened the pull request:
**Description**
The goal with this change is to have BaseModel which does not contain any database related code so that it can easily be extended to support other kinds of databases like NoSQL document based

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

najdanovicivan commented:
It's almost completely split the only issue so far is how to get rid of $primaryKey property

The problematic method is classToArray as it is defined as static method and $this->primaryKey is passed on each call. 

I’d suggest we change this method not to be static. Then it can be overridden inside Model class to handle primary while we still keep the conversion in base class.  

If we still need it for backwards compatibility reasons it can always be implemented via __callStatic in the Model.

najdanovicivan commented:
Another thought. I've used `db` to prefix methods in Model Class. When I take a look at BaseBuilder I see that `_` is used so it might be a good idea to use it in model class as well

mostafakhudair commented on system/BaseModel.php:
```diff
@@ -0,0 +1,1847 @@
+			$eventData = $this->trigger('beforeUpdate', $eventData);
+		}
+
+		$result = $this->dbUpdate($id, $eventData['data']);
+
+		$eventData = [
+			'id'     => $id,
+			'data'   => $eventData['data'],
+			'result' => $result,
+		];
+
+		if ($this->tempAllowCallbacks)
+		{
+			$this->trigger('afterUpdate', $eventData);
+		}
+
+		$this->tempAllowCallbacks = $this->allowCallbacks;
+
+		return $result;
```
```suggestion
		$eventData = [
			'id'     => $id,
			'data'   => $eventData['data'],
			'result' => $this->dbUpdate($id, $eventData['data']),
		];

		if ($this->tempAllowCallbacks)
		{
			$this->trigger('afterUpdate', $eventData);
		}

		$this->tempAllowCallbacks = $this->allowCallbacks;

		return $eventData['result'];
```

mostafakhudair commented on system/BaseModel.php:
```diff
@@ -0,0 +1,1847 @@
+			$this->trigger('beforeDelete', $eventData);
+		}
+
+		$result = $this->dbDelete($id, $purge);
+
+		$eventData = [
+			'id'     => $id,
+			'purge'  => $purge,
+			'result' => $result,
+			'data'   => null,
+		];
+
+		if ($this->tempAllowCallbacks)
+		{
+			$this->trigger('afterDelete', $eventData);
+		}
+
+		$this->tempAllowCallbacks = $this->allowCallbacks;
+
+		return $result;
```
```suggestion
		$eventData = [
			'id'     => $id,
			'data'   => null,
			'purge'  => $purge,
			'result' => $this->dbDelete($id, $purge),
		];

		if ($this->tempAllowCallbacks)
		{
			$this->trigger('afterDelete', $eventData);
		}

		$this->tempAllowCallbacks = $this->allowCallbacks;

		return $eventData['result'];
```

najdanovicivan commented:
@mostafakhudair Thanks on these

mostafakhudair commented on system/BaseModel.php:
```diff
@@ -0,0 +1,1832 @@
+		}
+
+		$eventData = [
+			'data'      => $this->dbFirst();,
```
```suggestion
			'data'      => $this->dbFirst(),
```
My fault copy paste failure

najdanovicivan commented:
I went ahead and removed ClassToArray method in favour of new object method objectToArray() as this method beside conversion to array also converts the Time instances to strings. 

So I've split the method in two methods objectToRawArray which does only the conversion and then objectToArray() internally call ObjectToRawArray

This makes it easy to override it it SQL Model to always add the PrimaryKey. This makes it possible to remove the primary key now from the BaseModel.

paulbalandan reviewed (changes requested):
Can we change the abstract protected functions to not start with underscore? It was corrected with variables a while back and it was too late to do the same with methods as it will break apps. Moving forward I hope this would not be seen in new code as it adds no value.

Also, in accord with PSR12 and PSR2 please reorder `protected abstract function` to `abstract protected function`.

najdanovicivan commented:
@paulbalandan I've intentionally put the underscores to follow the same methodology from BaseBuilder. I don't know what's the best way to go with that. I'm completely fine with any of these. I just wanted to be more consistent with the existing framework code.

There are some CodeSniffer rules I don't like but those should be separate PR but let me just state them. I'd like to here opinions before I'm even to start working with the PR

- Private methods are fine to start with _ but protected are not. But we have existing protected methods starting with _ in the code
- Line Length I believe it's 120 characters. It's just to low limit I'd rather got with 200 or 250 
- @throws require Message which ends with dot (.) I don't think we even need forcing for the message for these
- This I'm not sure about but maybe no forcing @param message as well. There are a lot of methods missing those

paulbalandan commented:
> * Private methods are fine to start with _ but protected are not. But we have existing protected methods starting with _ in the code

I think it was mentioned before that no methods should start with `_` except for `_remap` to retain BC with CI3. The practice of prepending underscores to imply non-public visibility was a remnant from CI3. Since it is too late to rename such other methods, those were left as-is.

> * Line Length I believe it's 120 characters. It's just to low limit I'd rather got with 200 or 250

I'm not aware that codesniffer enforces line length but I do not object with yours either. After all, PSR12 does not enforce a hard limit on line length.

> * @throws require Message which ends with dot (.) I don't think we even need forcing for the message for these

I'm not aware of this either. But grammatically speaking this should be the case since error messages are sentences.

> * This I'm not sure about but maybe no forcing @param message as well. There are a lot of methods missing those

I'm not sure what you meant by this if either `@param string $name` or `@param string $name the name`. Either way `@param` and `@return` docblocks are needed even though appearing as unneeded because the API generator needs them.

paulbalandan commented:
My suggestion for replacement of the non-public methods is this:

From:
```php
public function find();
protected function _find();
```

To:
```php
public function find();
protected function doFind();
```

najdanovicivan commented:
@paulbalandan regarding CodeSniffer rules

@param string $name **comment**

This is required and have to start with capital letter @param does not enforce the dot at the end

@throws Exception **comment** 

This is required and have to start with capital letter also have to end with dot

What does not make sense is that param comment does not require dot while throws comment requires it

najdanovicivan commented:
So the issue is due to these 

We have these in ruleset.xml
```
    <rule ref="Squiz.Commenting.FunctionComment.ParamCommentFullStop">
        <severity>0</severity>
    </rule>
```

But we're missing
```
    <rule ref="Squiz.Commenting.FunctionComment.ThrowsNoFullStop">
        <severity>0</severity>
    </rule>
```

So we should either require full stop at the end for both or not require it anywhere

paulbalandan commented:
I'm not really that well-versed with codesniffer but here are my suggestions:

1. `@param string $name comment` - `comment` should not start with a capital and neither stop with a period
2. `@throws Exception comment` - comment is optional and need not to be enforced for all. But if ever it appears there, it should follow the same rules with `@param`
3. `@throws` as well as other phpdocs should be explicitly provided although it is superfluous because PHPDocumentor requires it to generate our API documentation. Without these, the compilation will error.

If there are corresponding rulesets for what I mentioned, please open a PR to the `coding-standards` repo.

najdanovicivan commented:
> * `@param string $name comment` - `comment` should not start with a capital and neither stop with a period

@paulbalandan there are 2 different options here

By should not does it mean if it starts with capital letter and ends with dot it should be thrown as a warning or it does not matter and the rule enforcing this should be disabled ?

paulbalandan commented:
> > * `@param string $name comment` - `comment` should not start with a capital and neither stop with a period
> 
> @paulbalandan there are 2 different options here
> 
> By should not does it mean if it starts with capital letter and ends with dot it should be thrown as a warning or it does not matter and the rule enforcing this should be disabled ?

If there are rules that enforce all small letters and no period then those should be used instead. If none, then current rules should be disabled.

paulbalandan commented on system/BaseModel.php:
```diff
@@ -0,0 +1,1825 @@
+use stdClass;
+
+/**
+ * Class Model
+ *
+ * The Model class provides a number of convenient features that
+ * makes working with a database table less painful.
+ *
+ * It will:
+ *      - automatically connect to database
+ *      - allow intermingling calls between db connection, the builder,
+ *          and methods in this class.
+ *      - simplifies pagination
+ *      - removes the need to use Result object directly in most cases
+ *      - allow specifying the return type (array, object, etc) with each call
+ *      - ensure validation is run against objects when saving items
```
I think this docblock needs to be revised since this `BaseModel` is DB-agnostic?

najdanovicivan replied:
Sure. A lot of comments have to be reworked.
 
I also noticed something else. 
$this->table is not used anywhere in BaseModel now so it should be moved to Model as well. And remove $this->table from all the comments

Also with all this. I'm noticing now how simple it will be to add support for Composite Keys. I'm not sure if this is to be implemented with this PR or should I make additional one for such purpose 

```
	protected $primaryKey = ['composite1','composite2'];
```

Then find can be called like this

```
      $model->find(['composite1_value', 'composite2_value'])
```

paulbalandan replied:
Composite primary keys should be for another PR and discussion. 👍

najdanovicivan commented:
https://github.com/codeigniter4/coding-standard/pull/31 I've submitted PR for the CodeSniffer Rules

MGatner commented:
I do think this is really good and important but I will need to look at it more closely. I have a NoSQL model stub I would like to try putting on `BaseModel` to test it out. https://github.com/tattersoftware/codeigniter4-firebase/blob/develop/src/Model.php

paulbalandan reviewed (commented):
I'm okay with this but I'll wait first for @MGatner 's review.

MGatner commented:
I have reviewed the code but I still have not had a chance to test this with a NoSQL instance or my own custom Model class. We want these deprecations to post to `4.1` and tests are all passing so I will approve this PR, but please note that this split may require some more work down the road.

MGatner commented:
@paulbalandan I think you've finished reviewing but your Review has not been updated. I will give it a day.

najdanovicivan commented:
@MGatner I'm mostly certain that it will require some additional work down the road. But this should be enough to get it going. I'm already having idea of doing a further split for BaseBuilderModel which will come useful when additional drivers are implemented using builder class

So in the end we'll have something like  BaseModel->BaseBuilderModel->Model but everything is still raw with all of these. I already have some PRs planed for BaseConnection, QueryInterface and Database toolbar collector but I intend on submitting those one we have MongoDB driver ready to be publicly released so that everything makes sense and it is possible to do unit tests
